### ISSUE TITLE
writeHead called with a statusMessage sets message contents as headers

### ISSUE THREAD

**Issue by exogen** [gustavnikolaj/hijackresponse#12]
Hi Gustav!

I'm seeing strange response headers when using `hijackresponse`, even with just a simple always-`unhijack` setup or a simple pass-through transform:

```
HTTP/1.1 200 OK
0: O
1: K
x-powered-by: Express
accept-ranges: bytes
access-control-allow-origin: *
content-type: text/html; charset=UTF-8
Content-Length: 548
etag: W/"202-1SxF0UlPnK3ouKs7FF8xugo6bZk"
Vary: Accept-Encoding
date: Tue, 25 Apr 2017 09:03:38 GMT
connection: close
```

Notice the strange `0` and `1` headers that have adopted the OK status. Without `hijackresponse`, these are not present. This seems like a bug, or could it be a bad interaction with other middleware?

**Comment by gustavnikolaj** [comment]
Hi @exogen. I haven't seen that symptom before and we're using this module extensively in high load production apps, but it is kind of black magic, so it's not at all impossible that you've found a bug. Can you provide a small example app that reproduces the problem? :-) I don't really now where to start digging into this problem.

**Comment by exogen** [comment]
I should note that the next middleware in the chain is `http-proxy-middleware` – I'm trying to transform the proxy's response. When I do this in production without the proxy setup I'm not seeing these extra headers. Will try to come up with a minimal repro.

**Comment by gustavnikolaj** [comment]
Just tried adding the smallest example I could think of to the project and check the output:

```js
var express = require('express');
var hijackResponse = require('./');

var app = express();

app.use(function (req, res, next) {
    hijackResponse(res, function (err, res) {
        if (err) {
            res.unhijack(); // Make the original res object work again
            return next(err);
        }

        // Don't hijack HTML responses:
        if (/^text\/html(?:;$)/.test(res.getHeader('Content-Type'))) {
            return res.unhijack();
        }

        res.setHeader('X-Hijacked', 'yes!');
        res.removeHeader('Content-Length');

        res.pipe(res);
    });
    // next() must be called explicitly, even when hijacking the response:
    next();
});

app.use(function (req, res, next) {
    res.status(200);
    res.end();
});

app.listen(3000);
```

```
$ curl -v http://localhost:3000/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< X-Powered-By: Express
< X-Hijacked: yes!
< Date: Tue, 25 Apr 2017 09:26:36 GMT
< Connection: keep-alive
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
```

Is the curl command similar to what you did to get the output in your example above?

What are the characteristics of the response that you're proxying? Would a response with `Content-Type: text/plain` with a body of `foo` be enough to mimic it?

**Comment by gustavnikolaj** [comment]
Just tried this:

```js
var express = require('express');
var app = express();

app.use(function (req, res, next) {
    res.status(200);
    res.set('Content-Type', 'text/plain');
    res.end('foo');
});

app.listen(3001);
```

```js
var express = require('express');
var hijackResponse = require('./');

var app = express();

app.use(function (req, res, next) {
    hijackResponse(res, function (err, res) {
        if (err) {
            res.unhijack(); // Make the original res object work again
            return next(err);
        }

        // Don't hijack HTML responses:
        if (/^text\/html(?:;$)/.test(res.getHeader('Content-Type'))) {
            return res.unhijack();
        }

        res.setHeader('X-Hijacked', 'yes!');
        res.removeHeader('Content-Length');

        res.pipe(res);
    });
    // next() must be called explicitly, even when hijacking the response:
    next();
});

app.use('/proxy', require('http-proxy-middleware')({ target: 'http://localhost:3001' }));

app.use(function (req, res, next) {
    res.status(200);
    res.end();
});

app.listen(3000);
```

```
$ curl -v localhost:3000/proxy
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET /proxy HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< x-powered-by: Express
< content-type: text/plain; charset=utf-8
< date: Tue, 25 Apr 2017 09:35:00 GMT
< connection: close
< transfer-encoding: chunked
< X-Hijacked: yes!
< 
* Closing connection 0
foo
```

Seems to be working correctly so far. Hope that you'll succeed in reproducing it :-)

**Comment by exogen** [comment]
Looks like basic `http-proxy-middleware` usage is enough to trigger it:

```
var express = require('express');
var proxy = require('http-proxy-middleware');
var hijackResponse = require('hijackresponse');

var app = express();

app.use(function (req, res, next) {
  hijackResponse(res, function (err, res) {
    res.unhijack();
    if (err) {
      return next(err);
    }
  });
  next();
});

app.use(proxy({
  target: 'http://www.example.org',
  changeOrigin: true
}));

app.listen(3000);
```

Results from cURL:

```
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.51.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< 0: O
< 1: K
< X-Powered-By: Express
< accept-ranges: bytes
< cache-control: max-age=604800
< content-type: text/html
< date: Tue, 25 Apr 2017 09:37:06 GMT
< etag: "359670651"
< expires: Tue, 02 May 2017 09:37:06 GMT
< last-modified: Fri, 09 Aug 2013 23:54:35 GMT
< server: ECS (pae/3796)
< vary: Accept-Encoding
< x-cache: HIT
< content-length: 1270
< connection: close
< 
(body snipped)
```

Without the `changeOrigin: true` setting, an even more fun result:

```
HTTP/1.1 404 Not Found
0: N
1: o
2: t
3: 
4: F
5: o
6: u
7: n
8: d
(rest snipped)
```

**Comment by gustavnikolaj** [comment]
Running your exact example locally gives me the following output:

```
curl localhost:3000 -v
* Rebuilt URL to: localhost:3000/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< X-Powered-By: Express
< cache-control: max-age=604800
< content-type: text/html
< date: Tue, 25 Apr 2017 09:40:18 GMT
< etag: "359670651+gzip+ident"
< expires: Tue, 02 May 2017 09:40:18 GMT
< last-modified: Fri, 09 Aug 2013 23:54:35 GMT
< server: ECS (ewr/15BD)
< vary: Accept-Encoding
< x-cache: HIT
< content-length: 1270
< connection: close
< 
<!doctype html>
[...]
```

**Comment by exogen** [comment]
Hmm, what version of Node are you testing it on? I'm testing with:

```console
$ node -v
v6.9.5
```

**Comment by gustavnikolaj** [comment]
I've been running my tests with the following versions:

- O/S: Ubuntu 16.04
- node: 0.10.38
- express: 4.13.3
- http-proxy-middleware: 0.17.4

**Comment by gustavnikolaj** [comment]
Success! Or... well :-) I reproduced it. Thanks!

```
 $ curl localhost:3000 -v
* Rebuilt URL to: localhost:3000/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< 0: O
< 1: K
< X-Powered-By: Express
< cache-control: max-age=604800
< content-type: text/html
< date: Tue, 25 Apr 2017 09:43:20 GMT
< etag: "359670651+gzip+ident"
< expires: Tue, 02 May 2017 09:43:20 GMT
< last-modified: Fri, 09 Aug 2013 23:54:35 GMT
< server: ECS (ewr/15BD)
< vary: Accept-Encoding
< x-cache: HIT
< content-length: 1270
< connection: close
< 
<!doctype html>
[...]
```

Will try to dig in and see if I can find out what changed.

**Comment by exogen** [comment]
Confirmed that I don't see them with Node 0.10, but do with the latest LTS version (6.10.2).

**Comment by exogen** [comment]
Thanks for the help @gustavnikolaj! I'll look into it as well.

**Comment by gustavnikolaj** [comment]
I can reproduce it with my own example as well. Happens on node 4.8.0 too. And 0.12. But seemingly only when using the `http-proxy-middleware` module.

**Comment by exogen** [comment]
Looks like the issue may be in the `writeHead` override:

```js
  res.writeHead = function (statusCode, headers) {
    // When I debug this, headers == 'OK'
    if (statusCode) {
      res.statusCode = statusCode
    }
    if (headers) {
      // So here we're looping over indices of 'OK'...
      for (var headerName in headers) {
        // And setting them as headers.
        res.setHeader(headerName, headers[headerName])
      }
    }
    res.writeHead = writeHead
    cb(null, hijackedResponse)
  }
```

**Comment by exogen** [comment]
There can be 3 arguments to `writeHead`, optionally a status message: https://nodejs.org/api/http.html#http_response_writehead_statuscode_statusmessage_headers

(Not sure as of which version.)

**Comment by gustavnikolaj** [comment]
Great find! I've added a test case that will fail on node versions higher than 10. I'll try to fix it asap.

**Comment by gustavnikolaj** [comment]
Thanks a lot for your help! Just adding another argument to writeHead indeed made it go away. I manually verified that the code still works on 0.10 (even though we don't support it anymore) so it leads me to believe that this breaking change in `res.writeHead` is being worked around in either core or express. I still added that fallback so that we still use the headers if they are being passed as the second argument. I'll tag and release a new version as soon as I have a green light on CI :-)

**Comment by gustavnikolaj** [comment]
Published in [v3.0.0 ](https://github.com/gustavnikolaj/hijackresponse/commit/e82a37c907edbfdbb89ce17b19906b06b47fffa5). Might not strictly be a major version bump, but better be safe than sorry I guess :-)

**Comment by exogen** [comment]
Thanks a ton @gustavnikolaj, I just tested it with my app and it works like a charm!

#13 totally fixes my use case, but I wonder if anyone is relying on full `statusMessage` support – by that I mean, `writeHead` now handles being passed a `statusMessage`, but it doesn't actually do anything with it (I think it should probably set [res.statusMessage](https://nodejs.org/api/http.html#http_response_statusmessage) just like it does with `statusCode`). You'd probably also then want a `statusMessage` getter/setter on `hijackedResponse`, just like `statusCode` has.

I'm happy to open a PR to that effect if you think that sounds right. 😃

**Comment by gustavnikolaj** [comment]
If you have the time and need for it sure :-) I don't know how many people actually rely on status messages, so I think I'd rather just do it when it's needed than eagerly implementing it. Does that make sense?

I agree that we would want it to be there - when this module was written `res.statusMessage` was not part of the core `http.ServerResponse` (IIRC it was added in 0.11) and I guess we would have added it if we wrote it now. I would just prefer not to rock the boat until we know there's a demand :-)
